src/mesh/mesh_refinement_ops.hpp

```diff
@@ -0,0 +1,245 @@
+  const Real xp = GetX<DIM>(coarse_coords, i + 1);
+  dxm = xc - xm;
+  dxp = xp - xc;
+  const Real fxm = GetX<DIM>(coords, fi);
+  const Real fxp = GetX<DIM>(coords, fi + 1);
+  dxfm = xc - fxm;
+  dxfp = fxp - xc;
+}
+KOKKOS_FORCEINLINE_FUNCTION
+Real GradMinMod(const Real ccval, const Real fm, const Real fp, const Real dxm,
+                const Real dxp) {
+  const Real gxm = (ccval - fm) / dxm;
+  const Real gxp = (fp - ccval) / dxp;
+  return 0.5 * (SIGN(gxm) + SIGN(gxp)) * std::min(std::abs(gxm), std::abs(gxp));
+}
+} // namespace util
+
+template <int DIM>
+struct RestrictCellAverage {
+
+  KOKKOS_FORCEINLINE_FUNCTION static void
+  Do(const int l, const int m, const int n, const int ck, const int cj, const int ci,
+     const IndexRange &ckb, const IndexRange &cjb, const IndexRange &cib,
+     const IndexRange &kb, const IndexRange &jb, const IndexRange &ib,
+     const Coordinates_t &coords, const Coordinates_t &coarse_coords,
+     const ParArray6D<Real> &coarse, const ParArray6D<Real> &fine) {
+    const int i = (ci - cib.s) * 2 + ib.s;
+    int j = jb.s;
+    if (DIM > 1) {
+      j = (cj - cjb.s) * 2 + jb.s;
+    }
+    int k = kb.s;
+    if (DIM > 2) {
+      k = (ck - ckb.s) * 2 + kb.s;
+    }
+    // JMM: If dimensionality is wrong, accesses are out of bounds. Only
+    // access cells if dimensionality is correct.
+    const Real vol000 = coords.Volume(k, j, i);
+    const Real vol001 = coords.Volume(k, j, i + 1);
+    const Real fine000 = fine(l, m, n, k, j, i);
+    const Real fine001 = fine(l, m, n, k, j, i + 1);
+    Real vol010, vol011, vol100, vol101, vol110, vol111;
+    Real fine010, fine011, fine100, fine101, fine110, fine111;
+    vol010 = vol011 = vol100 = vol101 = vol110 = vol111 = 0;
+    fine010 = fine011 = fine100 = fine101 = fine110 = fine111 = 0;
+    if (DIM > 1) { // TODO(c++17) make constexpr
+      vol010 = coords.Volume(k, j + 1, i);
+      vol011 = coords.Volume(k, j + 1, i + 1);
+      fine010 = fine(l, m, n, k, j + 1, i);
+      fine011 = fine(l, m, n, k, j + 1, i + 1);
+    }
+    if (DIM > 2) { // TODO(c++17) make constexpr
+      vol100 = coords.Volume(k + 1, j, i);
+      vol101 = coords.Volume(k + 1, j, i + 1);
+      vol110 = coords.Volume(k + 1, j + 1, i);
+      vol111 = coords.Volume(k + 1, j + 1, i + 1);
+      fine100 = fine(l, m, n, k + 1, j, i);
+      fine101 = fine(l, m, n, k + 1, j, i + 1);
+      fine110 = fine(l, m, n, k + 1, j + 1, i);
+      fine111 = fine(l, m, n, k + 1, j + 1, i + 1);
+    }
+    // KGF: add the off-centered quantities first to preserve FP
+    // symmetry
+    const Real tvol =
+        ((vol000 + vol010) + (vol001 + vol011)) + ((vol100 + vol110) + (vol101 + vol111));
+    coarse(l, m, n, ck, cj, ci) =
+        (((fine000 * vol000 + fine010 * vol010) + (fine001 * vol001 + fine011 * vol011)) +
+         ((fine100 * vol100 + fine110 * vol110) +
+          (fine101 * vol101 + fine111 * vol111))) /
+        tvol;
+  }
+};
```

Comment:
This is an example of a functor. You need one per kind of operation you want to do. The struct is templated on `int DIM` and has a `KOKKOS_INLINE_FUNCTION` `Do` which takes indices, index ranges, coords, and the `ParArray`s to operate on. The functor class is passed into the loop construct, but it doesn't have to be specialized to a particular `DIM`. The loop construct can do that for you.